File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
#define X first
#define Y second

using namespace std;
queue<pair<int, int>> q;
int dist[105][105];

int dx[4] = {1, 0, -1, 0};
int dy[4] = {0, 1, 0, -1};
int solution(vector<vector<int>> maps)
{
    int n = (int)maps.size();
    int m = (int)maps[0].size();
    
    for(int i = 0; i < n; i++){
        fill(dist[i], dist[i] + m, -1);
    }
    
    q.push({0, 0});
    dist[0][0] = 1;
    
    while(!q.empty()){
        auto cur = q.front(); q.pop();

        for(int dir = 0; dir < 4; dir++){
            int nx = cur.X + dx[dir];
            int ny = cur.Y + dy[dir];

            if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
            if(dist[nx][ny] >= 0 || maps[nx][ny] == 0) continue;

            dist[nx][ny] = dist[cur.X][cur.Y] + 1;
            q.push({nx, ny});
        }
    }
    
    
    return dist[n - 1][m - 1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};
    static Queue<int[]> q = new ArrayDeque<>();
    static int[][] dist;
    static int n, m;
    
    public int solution(int[][] maps) {    
        n = maps.length;
        m = maps[0].length;
        
        dist = new int[n][m];
        
        for(int i = 0; i < n; i++)
            Arrays.fill(dist[i], -1);
        
        q.offer(new int[] {0, 0});
        dist[0][0] = 1;
        
        while(!q.isEmpty()){
            int[] cur = q.poll();
            
            int x = cur[0];
            int y = cur[1];
            
            for(int dir = 0; dir < 4; dir++){
                int nx = x + dx[dir];
                int ny = y + dy[dir];
                
                if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if(maps[nx][ny] == 0 || dist[nx][ny] >= 0) continue;
                
                dist[nx][ny] = dist[x][y] + 1;
                q.offer(new int[] {nx, ny});
            }
        }
        
        return dist[n - 1][m - 1];
    }
}